engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java

```diff
@@ -110,20 +110,40 @@ private void logSummary(final List<MigrationTask> migrationTasks) {
 
   private boolean handleMigrationTask(
       final MigrationTask migrationTask, final int index, final int total) {
+    // The migration is already finished. We don't need to run it again.
+    if (processingState.getMigrationState().isMigrationFinished(migrationTask.getIdentifier())) {
+      logMigrationTaskAlreadyExecuted(migrationTask, index, total);
+      return false;
+    }
+
     if (migrationTask.needsToRun(processingState)) {
       try {
         currentMigration = migrationTask;
         runMigration(migrationTask, index, total);
+        processingState.getMigrationState().markMigrationFinished(migrationTask.getIdentifier());
       } finally {
         currentMigration = null;
       }
       return true;
     } else {
       logMigrationSkipped(migrationTask, index, total);
+      processingState.getMigrationState().markMigrationFinished(migrationTask.getIdentifier());
       return false;
     }
```

Comment:
🔧 Currently the logic is a bit mixed. We could simplify it using an early return, i.e. invert the if statement that checks whether the task needs to run

Then the logic is:
- if already finished, no need to run it again
- if it doesn't have to run, no need to run it again but marking as finished
- else run it and mark it as finished at the end